lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -401,6 +401,17 @@ struct CircuitLoweringState {
     macroDeclNames.insert(name);
   }
 
+  /// The list of fragments on which the modules rely. Must be set outside the
+  /// parallelized module lowering since module type reads access it.
+  DenseMap<hw::HWModuleOp, SetVector<Attribute>> fragments;
+  std::mutex fragmentsMutex;
+
+  void addFragment(hw::HWModuleOp module, StringRef fragment) {
+    std::unique_lock<std::mutex> lock(fragmentsMutex);
+    fragments[module].insert(FlatSymbolRefAttr::get(
+        StringAttr::get(circuitOp.getContext(), fragment)));
```

Comment:
Nit: just using the Context and StringRef `get` method is likely cleaner than wrapping these in a `StringAttr` given that they use the same arguments:
```suggestion
    fragments[module].insert(FlatSymbolRefAttr::get(
        circuitOp.getContext(), fragment));
```